File: src/java1/algorithms/hashmap/pathCrossing/PathCrossing.java
package java1.algorithms.hashmap.pathCrossing;

import java.util.HashSet;
import java.util.Set;

public class PathCrossing {
    private static boolean isPathCrossing(String path) {
        Set<String> visited = new HashSet<>();
        int x=0, y = 0;

        visited.add(x+","+y);

        for (char move : path.toCharArray()) {
            if(move == 'N') y++;
            if(move == 'S') y--;
            if(move == 'E') x++;
            if(move == 'W') x--;

            if(visited.contains(x+","+y)) {
                return true;
            }
    
            visited.add(x+","+y);
        }

        return false;
    }

    public static void main(String[] args) {
        String path1="NESE";
        String path2="NEWS";

        System.out.println(isPathCrossing(path1));
        System.out.println(isPathCrossing(path2));
    }
}

File: src/javascript/algorithms/hashtable/12.pathCrossing/pathCrossing.js
function isPathCrossing(path){
    const visited = new Set();

    let x = 0, y = 0;
    visited.add(`${x},${y}`);

    const directions = {
        'N': [0, 1],
        'S': [0, -1],
        'E': [1, 0],
        'W': [-1, 0]
    }

    for (const move of path) {
        x+= directions[move][0],
        y+= directions[move][1]

        if(visited.has(`${x},${y}`)) {
            return true;
        }

        visited.add(`${x},${y}`);
    }

    return false;
}

const path1="NESE";
const path2="NEWS";

console.log(isPathCrossing(path1));
console.log(isPathCrossing(path2));